### PR TITLE
Import configuration when migrating a cluster from Ansible

### PR DESCRIPTION
**Pull request by yuzhibotao**
### What problem does this PR solve? 
From ansible to tiup ，there is no configuration parameter for the import action.

### What is changed and how it works?

1. If the monitored_server is configured with deploy_dir and they are all the same, import deploy_dir into monitored_servers.
2.  Add information about global parameters and component parameters of TiDB PD TiKV .
After backing up the configuration files, parse the configuration files for each component, extract the same parameters into Server_Configs and put different parameters into the service's own Config.

### Check List 

Tests 
 - Manual test (add detailed scripts or steps below)
 
[old_meta.txt](https://github.com/pingcap/tiup/files/5192422/old_meta.txt)
[new_meta.txt](https://github.com/pingcap/tiup/files/5192428/new_meta.txt)

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

Release notes:

```release-note
NONE
```

### CODE REVIEW

**Comment by lonng** [comment]
@yuzhibotao Please add some tests and format your code.
@AstroProfundis PTAL

File: pkg/cluster/ansible/import.go [hunk 1]
```diff
@@ -113,3 +137,237 @@ func SSHKeyPath() string {
...
+			}
+		}
+	}
+	return srvConfigs, nil
+}
+
+
+// Load config files to clusterMeta ,just include tidbservers, tikvservers and pdservers
+func LoadConfig(clsName string ,cls *spec.ClusterMeta) error {
+	prefixkey := ""
```
Comment (AstroProfundis) [inline pkg/cluster/ansible/import.go:212]: `prefixKey`

File: pkg/cluster/ansible/import.go [hunk 2]
```diff
@@ -113,3 +137,237 @@ func SSHKeyPath() string {
...
+	var err error
+	Configs :=  make(map[string]interface{})
+	gConfigs := make(map[string]interface{})
+	initflag := false
+	for i := 0 ; i < len(cls.Topology.TiDBServers); i++ {
```
Comment (AstroProfundis) [inline pkg/cluster/ansible/import.go:217]: Use `for i, srv := range cls.Topology.TiDBServers {...}` and then you can use `srv.ComponentName()` to replace `prefixkey`.

File: pkg/cluster/ansible/import.go [hunk 3]
```diff
@@ -113,3 +137,237 @@ func SSHKeyPath() string {
...
+	}
+	return srvConfigs, nil
+}
+
+
+// Load config files to clusterMeta ,just include tidbservers, tikvservers and pdservers
+func LoadConfig(clsName string ,cls *spec.ClusterMeta) error {
+	prefixkey := ""
+	var err error
+	Configs :=  make(map[string]interface{})
```
Comment (AstroProfundis) [inline pkg/cluster/ansible/import.go:214]: no need to export the variable

File: pkg/cluster/ansible/import.go [hunk 4]
```diff
@@ -113,3 +137,237 @@ func SSHKeyPath() string {
...
+		prefixkey = "tidb"
+		srv := cls.Topology.TiDBServers[i]
+		fname := spec.ClusterPath(clsName,spec.AnsibleImportedConfigPath,fmt.Sprintf("%s-%s-%d.toml",prefixkey,srv.Host, srv.Port))
+
+		Configs , err  = ParseConfigFile(fname)
+		if initflag == false {
+			gConfigs = Configs
+			initflag = true
+		}
+		// make Configs containts global Config variables
```
Comment (AstroProfundis) [inline pkg/cluster/ansible/import.go:227]: We have `Merge2Toml()` in `pkg/cluster/spec/server_config.go` that could be helpful for this purpose.

File: pkg/cluster/ansible/import.go [hunk 5]
```diff
@@ -113,3 +137,237 @@ func SSHKeyPath() string {
...
+		for k, _ := range gConfigs {
+			if Configs[k] == nil && gConfigs[k] != nil {
+				Configs[k] = k
+			}
+		}
+
+		for k, v := range  Configs {
+            //make sure variables should be put in Server Configs or Global Configs
```
Comment (AstroProfundis) [inline pkg/cluster/ansible/import.go:235]: Could you explain more detail about what do this code block do and how it works?

File: pkg/cluster/ansible/import.go [hunk 6]
```diff
@@ -113,3 +137,237 @@ func SSHKeyPath() string {
...
+				}
+				delete(cls.Topology.ServerConfigs.TiDB,k)
+			}
+		}
+	}
+
+	Configs =  make(map[string]interface{})
+	gConfigs = make(map[string]interface{})
+	initflag = false
+	for i := 0 ; i < len(cls.Topology.TiKVServers); i++ {
```
Comment (AstroProfundis) [inline pkg/cluster/ansible/import.go:272]: ditto

File: pkg/cluster/spec/spec_manager.go [hunk 7]
```diff
@@ -99,7 +100,13 @@ func (s *SpecManager) SaveMeta(clusterName string, meta Metadata) error {
 	}
 
 	data, err := yaml.Marshal(meta)
-	if err != nil {
+
+	// remove quotes
```
Comment (AstroProfundis) [inline pkg/cluster/spec/spec_manager.go:104]: `Marshal()` converts the object to **valid** YAML bytes, so it's not a good idea to replace anything after it.

If the processing methods are all proper when building the `meta` object, it should be no invalid quote chars in the object.

**Review by AstroProfundis — CHANGES_REQUESTED** [review]
You may use `go fmt` to format your code, and use `make check && make test` for basic code formatting and unit test checks.

File: pkg/cluster/ansible/import.go [hunk 8]
```diff
@@ -100,6 +113,17 @@ func parseInventoryFile(invFile io.Reader) (string, *spec.ClusterMeta, *aini.Inv
 	} else {
 		return "", nil, inventory, errors.New("no available host in the inventory file")
 	}
+
+	// get location_labels variables
+	if location_labels, ok := inventory.Groups["pd_servers"]; ok && len(location_labels.Vars["location_labels"]) > 0 {
```
[the remark this answers is not in the document]
Comment (yuzhibotao) [reply pkg/cluster/ansible/import.go:118]: Yes, but after importing, we will use edit-config to view the cluster topology and other detail information. The backup file cannot be displayed here, which is not friendly to dba.

File: pkg/cluster/ansible/import.go [hunk 9]
```diff
@@ -113,3 +137,237 @@ func SSHKeyPath() string {
 
 	return fmt.Sprintf("%s/.ssh/id_rsa", homeDir)
 }
+
+// CheckVarSame compare the variable in all Hosts same or not
+func CheckVarSame (key string, hosts map[string]*aini.Host) ([]string) {
```
[the remark this answers is not in the document]
Comment (yuzhibotao) [reply pkg/cluster/ansible/import.go:142]: Yes no need , I wanted to reuse this function for parameter comparison, but I gave up later.

File: pkg/cluster/ansible/import.go [hunk 10]
```diff
@@ -113,3 +137,111 @@ func SSHKeyPath() string {
...
+	for _, config := range locals {
+		for k := range global {
+			delete(config, k)
+		}
+	}
+
+	return
+}
+
+// LoadConfig files to clusterMeta, just include tidbservers, tikvservers and pdservers
```
Comment (AstroProfundis) [inline pkg/cluster/ansible/import.go:193]: What about pump and drainer?

File: pkg/cluster/ansible/import_test.go [hunk 11]
```diff
@@ -147,3 +143,61 @@ func sortClusterMeta(clsMeta *spec.ClusterMeta) {
...
+		c.Assert(err, IsNil)
+		c.Assert(m["x"], IsNil)
+		c.Assert(m["a"], Equals, true)
+		c.Assert(m["b.c"], Equals, int64(1))
+		c.Assert(m["b.d"], Equals, "\"")
+	})
+}
+
+func (s *ansSuite) TestDiffConfig(c *C) {
+	global, locals := diffConfigs([]map[string]interface{}{
```
Comment (AstroProfundis) [inline pkg/cluster/ansible/import_test.go:182]: should add cases for multi-level keys, e.g., `foo.bar: 1`.

File: pkg/cluster/spec/tikv.go [hunk 12]
```diff
@@ -125,7 +125,15 @@ func (s TiKVSpec) Labels() (map[string]string, error) {
 	lbs := make(map[string]string)
 
 	if serverLbs := GetValueFromPath(s.Config, "server.labels"); serverLbs != nil {
-		for k, v := range serverLbs.(map[interface{}]interface{}) {
+		m := map[interface{}]interface{}{}
+		if sm, ok := serverLbs.(map[string]interface{}); ok {
```
Comment (AstroProfundis) [inline pkg/cluster/spec/tikv.go:129]: `serverLbs` -> `serverLabels`

File: pkg/set/string_set.go [hunk 13]
```diff
@@ -13,6 +13,82 @@
 
 package set
 
+// AnySet is a set stores interface{}
+type AnySet struct {
```
Comment (AstroProfundis) [inline pkg/set/string_set.go:17]: better split it to a dedicated file, e.g., `any_set.go`

**Review by AstroProfundis — COMMENTED** [review]
Rest LGTM

**Review by AstroProfundis — APPROVED** [review]

**Pull request merged**